ledger/src/shred/traits.rs

```diff
@@ -0,0 +1,93 @@
+use {
+    crate::shred::{CodingShredHeader, DataShredHeader, Error, ShredCommonHeader, ShredFlags},
+    solana_sdk::{clock::Slot, signature::Signature},
+};
+
+pub(super) trait Shred: Sized {
+    fn from_payload(shred: Vec<u8>) -> Result<Self, Error>;
+    fn common_header(&self) -> &ShredCommonHeader;
+    fn sanitize(&self) -> Result<(), Error>;
+
+    fn set_signature(&mut self, signature: Signature);
+
+    fn payload(&self) -> &Vec<u8>;
+    fn into_payload(self) -> Vec<u8>;
+
+    // Returns the shard index within the erasure coding set.
+    fn erasure_shard_index(&self) -> Option<usize>;
+    // Returns the portion of the shred's payload which is erasure coded.
+    fn erasure_shard(self) -> Result<Vec<u8>, Error>;
+    // Like Shred::erasure_shard but returning a slice.
+    fn erasure_shard_as_slice(&self) -> Result<&[u8], Error>;
+
+    // Portion of the payload which is signed.
+    fn signed_payload(&self) -> &[u8];
+
+    // Possibly zero pads bytes stored in blockstore.
+    fn resize_stored_shred(shred: Vec<u8>) -> Result<Vec<u8>, Error>;
+
+    // Only for tests.
```

Comment:
can `#[cfg(test)]` be used here?

Reply:
These two functions are invoked outside of this crate; e.g:
https://github.com/solana-labs/solana/blob/24122768d/core/src/window_service.rs#L818
https://github.com/solana-labs/solana/blob/24122768d/core/src/repair_weighted_traversal.rs#L282

If we make these `#[cfg(test)]` then the functions will not be available outside the ledger crate.
`#[cfg(test)]` only works for tests within the same crate.

Reply:
ah, right. I know this predates this PR, but would it be useful to prefix these functions wtih `testing_` or something to indicate the intent? feel free to disregard if this is going to cause churn between the refactoring PRs.

Reply:
yeah, probably at some point once current changes are done, it would be good to rename these methods.